### PR TITLE
[primTorch] Implement batch, group, and instance norm references

### PR DESCRIPTION
**Pull request by rdspring1**
Add References:

1. batch norm
2. group norm
3. instance norm

Depends on:

- expand https://github.com/pytorch/pytorch/pull/79820
- repeat https://github.com/pytorch/pytorch/pull/81374

cc @ezyang @mruberry @ngimel @Lezcano @fdrocha @peterbell10

### CODE REVIEW

File: torch/_refs/__init__.py [hunk 1]
```diff
@@ -2139,6 +2142,61 @@ def flipud(a: TensorLikeType) -> TensorLikeType:
     return flip(a, (0,))
 
 
+def _repeat_if_defined(a: Optional[Tensor], sizes: int):
+    if a is not None:
+        return a.repeat(sizes)
+    return a
+
+
+@register_decomposition(torch.ops.aten.instance_norm)
```
Comment (Chillee) [inline torch/_refs/__init__.py:2151]: I think in this case, we don't actually need to register a decomposition since it's already a `CompositeImplicitAutograd` op (and so is automatically decomposed in the `__torch_dispatch__` level).

Comment (IvanYashchuk) [reply torch/_refs/__init__.py:2218]: Is there a test that complains when a decomposition for CompositeImplicitAutograd aten op is registered? Would this decomposition be used somewhere or ignored?

Comment (IvanYashchuk) [reply torch/_refs/__init__.py]: @Chillee, could you please take a look at the above questions?

Comment (Chillee) [reply torch/_refs/__init__.py]: I think it mostly won't be used/tested anywhere. I think there's a test that complains here: https://github.com/pytorch/pytorch/blob/master/test/test_ops.py#L1598

No big deal though imo.

File: torch/_refs/__init__.py [hunk 2]
```diff
@@ -2426,14 +2490,186 @@ def _normalize(
...
         mean (Tensor): mean of the tensor along norm_dims.
         rstd (Tensor): 1/std of the tensor along norm_dims.
+        unbiased_var (Tensor): variance of the tensor along norm_dims.
     """
     computation_dtype = utils.get_computation_dtype(a.dtype)
     a_acc = _maybe_convert_to_dtype(a, computation_dtype)
     assert isinstance(a_acc, TensorLike)  # to avoid mypy error for var_mean
-    biased_var, mean = var_mean(a_acc, dim=norm_dims, unbiased=False, keepdim=True)
-    rstd = torch.rsqrt(biased_var + eps)
+    unbiased_var, mean = var_mean(a_acc, dim=norm_dims, unbiased=False, keepdim=True)
```
Comment (IvanYashchuk) [inline torch/_refs/__init__.py:2498]: `unbiased` is set to `False` in the call, so renaming `biased_var` -> `unbiased_var` is incorrect

File: torch/_refs/__init__.py [hunk 3]
```diff
@@ -2426,14 +2490,186 @@ def _normalize(
...
+    rstd = torch.rsqrt(unbiased_var + eps)
     out = (a - mean) * rstd
-    return out, mean, rstd
+    return out, mean, rstd, unbiased_var
+
+
+# squeeze backwards from ndims-1 to 0
+def _squeeze_multiple(a: Tensor, dims: List[int]) -> Tensor:
```
Comment (IvanYashchuk) [inline torch/_refs/__init__.py:2505]: `prims.squeeze` squeezes multiple times, this helper function is not needed.

Comment (rdspring1) [reply torch/_refs/__init__.py:2505]: I thought we were supposed to use the `torch` functions instead of directly calling `prims` / `refs`.

Comment (IvanYashchuk) [reply torch/_refs/__init__.py:2505]: I can't dig up the message right now, but using torch functions is not strict. We can use prims/refs if it offers new needed functionality like in this case squeezing multiple times.

File: torch/_refs/__init__.py [hunk 4]
```diff
@@ -2642,6 +2645,61 @@ def flipud(a: TensorLikeType) -> TensorLikeType:
     return flip(a, (0,))
 
 
+def _repeat_if_defined(a: Optional[Tensor], sizes: int):
```
Comment (mruberry) [inline torch/_refs/__init__.py:2648]: Add a comment for this (private) function

File: aten/src/ATen/native/Normalization.cpp [hunk 5]
```diff
@@ -155,6 +155,8 @@ std::tuple<Tensor,Tensor> batch_norm_cpu_update_stats_template(
     double momentum, double eps) {
 
   using accscalar_t = at::acc_type<scalar_t, false>;
+  TORCH_CHECK(input.dim() >= 1,
```
Comment (mruberry) [inline aten/src/ATen/native/Normalization.cpp:158]: Since `batch_norm_cpu_update_stats_template` is not a user-facing function this should be an internal assertion -- maybe the check could occur earlier?

File: aten/src/ATen/native/Normalization.cpp [hunk 6]
```diff
@@ -455,8 +457,12 @@ std::tuple<Tensor, Tensor, Tensor, Tensor, int64_t> _batch_norm_impl_index(
 
     // don't return view of input, don't return empty tensor because it will break gradient chain
     auto out = input.clone();
-    if (weight.defined()) out = out * weight[0];
-    if (bias.defined()) out = out + bias[0];
+    if (weight.defined() && weight.numel() > 0) {
```
Comment (mruberry) [inline aten/src/ATen/native/Normalization.cpp:460]: This will need a comment for what the semantics are when `weight` and/or `bias` have no elements

File: aten/src/ATen/native/Normalization.cpp [hunk 7]
```diff
@@ -668,6 +674,18 @@ std::tuple<Tensor, Tensor, Tensor> batch_norm_cpu(const Tensor& self, const c10:
 
   checkBackend("batch_norm_cpu", {self, weight, bias, running_mean, running_var}, Backend::CPU);
 
+  // Handle size-0 inputs
+  if (self.numel() == 0) {
+    auto num_features = self.sizes()[1];
+    auto options = self.options().dtype(
+        at::toAccumulateType(self.scalar_type(), /*is_cuda=*/self.is_cuda()));
+    auto save_mean = at::empty({num_features}, options);
```
Comment (mruberry) [inline aten/src/ATen/native/Normalization.cpp:682]: Is it really OK to return tensors with random values here?

File: torch/testing/_internal/common_methods_invocations.py [hunk 8]
```diff
@@ -401,8 +401,6 @@ def sample_inputs_batch_norm(op_info, device, dtype, requires_grad, **kwargs):
         ((S, S, S), {'training': True, 'momentum': 0.5, 'eps': 0.6}),
         ((3, 2, 4), {'training': False, 'momentum': -1.2}),
         ((3, 1), {'training': True, 'momentum': 0.0}),
-        ((0,), {'training': True}),
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:404]: Should these be added to instance and groupnorm sample inputs, too?

**Review by mruberry — COMMENTED** [review]
Updates from offline review:

- let's separate the changes to existing operators and ensure those are tested properly
- also review https://github.com/pytorch/pytorch/issues/85960 and see if that should be part of the PR changing the ATen operations
- after the C++ changes to existing ATen operators, let's revisit the references
